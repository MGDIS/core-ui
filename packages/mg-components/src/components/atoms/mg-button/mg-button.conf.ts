import type { Variant, Size } from '../../../types';

/**
 * List of all possibles sizes
 */
export const sizes = ['medium', 'large'] as const;

/**
 * SizeType type from button sizes
 */
export type SizeType = Size & (typeof sizes)[number];

/**
 * List of all possibles variants
 */
export const variants = ['primary', 'secondary', 'danger', 'danger-alt', 'info', 'flat', 'success', 'link'] as const;

/**
 * VariantType type from button variants
 */
export type VariantType = Variant & (typeof variants)[number];

/**
 * List of all possibles button types
 */
export const buttonTypes = ['button', 'submit', 'reset'] as const;

/**
 * ButtonType type from buttonTypes
 */
export type ButtonType = (typeof buttonTypes)[number];
