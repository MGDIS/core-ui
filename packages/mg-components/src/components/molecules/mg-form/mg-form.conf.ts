/**
 * List of all possibles roles
 */
export const roles = ['form', 'search', 'none', 'presentation'] as const;

/**
 * Aria Role type from roles
 */
export type AriaRoleType = (typeof roles)[number];
