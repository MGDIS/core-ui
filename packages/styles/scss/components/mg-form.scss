@import 'form';

/*!
 * @prop --mg-c-form-inputs-title-width: Defines the width of the input title. Fallback on `calc(28 * var(--mg-b-size-8))`.
 */

.mg-c-form {
  --mg-c-input-title-width: var(--mg-c-form-inputs-title-width, calc(28 * var(--mg-b-size-8))); // 224px
  --mg-c-input-margin-bottom: var(--mg-b-size-16);
  --mg-c-input-title-horizontal-space: var(--mg-b-size-32);

  &.mg-c-form--all-required {
    --mg-c-input-title-display: none;
  }

  &.mg-c-form--label-on-top {
    --mg-c-input-margin-bottom: var(--mg-b-size-24);
  }

  & > p {
    font-size: 1.2rem;
  }
}
