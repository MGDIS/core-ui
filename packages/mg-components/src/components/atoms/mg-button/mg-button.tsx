import { Component, Element, h, Prop, State, Watch, Host, EventEmitter, Event } from '@stencil/core';
import { variants, VariantType, ButtonType, SizeType, sizes } from './mg-button.conf';
import { ClassList, isValidString, nextTick, toString } from '@mgdis/stencil-helpers';

/**
 * @slot - Button content
 */
@Component({
  tag: 'mg-button',
  styleUrl: '../../../../node_modules/@mgdis/styles/dist/components/mg-button.css',
  shadow: true,
})
export class MgButton {
  /************
   * Internal *
   ************/

  private onClickElementFn = null;
  private readonly classDisabled = 'mg-c-button--disabled';
  private readonly classLoading = 'mg-c-button--loading';
  private readonly classFullWidth = 'mg-c-button--full-width';

  /**************
   * Decorators *
   **************/

  /**
   * Get component DOM element
   */
  @Element() element: HTMLMgButtonElement;

  /**
   * Define button variant
   */
  @Prop({ mutable: true }) variant: VariantType = variants[0]; // Primary
  @Watch('variant')
  validateVariant(newValue: VariantType, oldValue?: VariantType): void {
    // validate new value
    if (!variants.includes(newValue)) {
      if ((newValue as string) === '') {
        // Reactive framework like VUE render with `""` when prop is update to undefined
        this.variant = variants[0];
        return;
      }
      throw new Error(`<mg-button> prop "variant" must be one of: ${variants.join(', ')}. Passed value: ${toString(newValue)}.`);
    } else {
      if (oldValue !== undefined) {
        this.classCollection.delete(`mg-c-button--${oldValue}`);
      }
      this.classCollection.add(`mg-c-button--${newValue}`);
    }
  }

  /**
   * aria-label
   * In case button text is not explicit enough
   */
  @Prop() label?: string;

  /**
   * Define button type
   */
  @Prop() type?: ButtonType;

  /**
   * Define button size
   */
  @Prop() size: SizeType = sizes[0]; // medium
  @Watch('size')
  validateSize(newValue: MgButton['size'], oldValue?: MgButton['size']): void {
    // validate new value
    if (!sizes.includes(newValue)) {
      throw new Error(`<mg-button> prop "size" must be one of: ${sizes.join(', ')}. Passed value: ${toString(newValue)}.`);
    } else {
      if (oldValue !== undefined) {
        this.classCollection.delete(`mg-c-button--${oldValue}`);
      }
      this.classCollection.add(`mg-c-button--${newValue}`);
      this.setIconSize();
    }
  }

  /**
   * Set button to full-width
   */
  @Prop() fullWidth = false;
  @Watch('fullWidth')
  validateFullWidth(newValue: MgButton['fullWidth']): void {
    if (newValue && this.isIcon) {
      throw new Error('<mg-button> prop "fullWidth" cannot be used with prop "isIcon".');
    } else if (newValue) {
      this.classCollection.add(this.classFullWidth);
    } else {
      this.classCollection.delete(this.classFullWidth);
    }
  }

  /**
   * Define form id to attach button with.
   * If this attribute is not set, the `<button>` is associated with its ancestor `<form>` element.
   */
  @Prop() form?: string;

  /**
   * Disable button
   */
  @Prop({ mutable: true }) disabled?: boolean;
  @Watch('disabled')
  disabledHandler(isDisabled: MgButton['disabled']): void {
    // Remove loading when enable
    // Will be set back onclick
    if (!isDisabled && this.disableOnClick) {
      this.loading = false;
    }
    // Manage if onclick
    this.element.onclick = isDisabled ? undefined : this.onClickElementFn;

    // apply style
    if (isDisabled) {
      this.classCollection.add(this.classDisabled);
    } else {
      this.classCollection.delete(this.classDisabled);
    }

    // emit value update
    this.disabledChange.emit(isDisabled);
  }
  /**
   * Define if button is round.
   * Used for icon button.
   */
  @Prop() isIcon = false;
  @Watch('isIcon')
  watchIsIcon(): void {
    this.setIconSize();
  }

  /**
   * Option to set input disable on click, in order to prevent multi-click.
   * Parent component have to remove the attribute 'disabled' when the process ends.
   */
  @Prop() disableOnClick = false;

  /**
   * Define if button is loading, default to false.
   * Trigger when button is clicked or key-up ['enter', 'space], then value change to true.
   * It's required to reset to false when action/promise in parent is done to stop the loading state
   */
  @State() loading = false;
  @Watch('loading')
  loadingHandler(newValue: MgButton['loading']): void {
    // we add loading style if it newvalue is true else we remove it
    if (newValue) {
      this.classCollection.add(this.classLoading);
    } else {
      this.classCollection.delete(this.classLoading);
    }
  }

  /**
   * Component classes
   */
  @State() classCollection: ClassList = new ClassList(['mg-c-button']);

  /**
   * Emmited event when disabled change
   */
  @Event({ eventName: 'disabled-change' }) disabledChange: EventEmitter<HTMLMgButtonElement['disabled']>;

  /**
   * Trigger actions onClick event
   * @param event - click event
   */
  private handleClick = (event: MouseEvent): void => {
    if (this.disabled) event.stopPropagation();
    // Used to prevent multi-click.
    else if (this.disableOnClick) {
      this.loading = true;
      this.disabled = true;
    }
  };

  /**
   * Handle onKeydown event
   * @param event - keyboard event
   */
  private handleKeydown = (event: KeyboardEvent): void => {
    if (!this.disabled && event.key === ' ') {
      event.preventDefault();
    } else if (!this.disabled && ['Enter', 'NumpadEnter', 'Space'].includes(event.key)) {
      event.preventDefault();
      this.element.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    }
  };

  /**
   * Handle onKeyup event
   * @param event - keyboard event
   */
  private handleKeyup = (event: KeyboardEvent): void => {
    if (!this.disabled && event.key === ' ') {
      event.preventDefault();
      this.element.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    }
  };

  /**
   * Set child `<mg-icon>` element `size` prop to follow the current `size` prop
   */
  private setIconSize = (): void => {
    if (this.size === 'medium') {
      this.element.querySelector('mg-icon')?.removeAttribute('size');
    } else if (this.isIcon) {
      this.element.querySelector('mg-icon')?.setAttribute('size', this.size);
    }
  };

  /**
   * Check if props are well configured on init
   */
  componentWillLoad(): void {
    this.validateVariant(this.variant);
    this.validateFullWidth(this.fullWidth);
    this.validateSize(this.size);
    this.watchIsIcon();
    if (this.isIcon) {
      this.classCollection.add(`mg-c-button--icon`);
      if (!isValidString(this.label)) {
        throw new Error(`<mg-button> prop "label" is mandatory when prop "isIcon" is set to true.`);
      }
    }
    // Store the onclick fn
    this.onClickElementFn = this.element.onclick;
    this.disabledHandler(this.disabled);
  }

  /**
   * Add listners
   */
  componentDidLoad(): void {
    nextTick(() => {
      const closestForm = this.element.closest('form') || this.element.closest('mg-form')?.shadowRoot.querySelector('form');
      // submit buttons should trigger form submition;
      if (closestForm && ['submit', undefined].includes(this.type)) {
        this.element.addEventListener('click', () => {
          closestForm.dispatchEvent(new SubmitEvent('submit', { bubbles: true, cancelable: true }));
        });
      }
    });
  }

  /**
   * Render component
   * @returns html element
   */
  render(): HTMLElement {
    return (
      <Host
        role="button"
        tabIndex={0}
        type={this.type}
        form={this.form}
        full-width={this.fullWidth}
        aria-label={this.label}
        aria-disabled={this.disabled?.toString()}
        onClick={this.handleClick}
        onKeyup={this.handleKeyup}
        onKeydown={this.handleKeydown}
      >
        <div class={this.classCollection.join()}>
          {this.loading && <mg-icon icon="loader" spin></mg-icon>}
          <div class="mg-c-button__content">
            <slot></slot>
          </div>
        </div>
      </Host>
    );
  }
}
