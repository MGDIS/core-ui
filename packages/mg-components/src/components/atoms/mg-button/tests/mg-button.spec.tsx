import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { MgButton } from '../mg-button';
import { MgForm } from '../../../molecules/mg-form/mg-form';
import { variants, buttonTypes, sizes } from '../mg-button.conf';
import { setupMutationObserverMock, setupSubmitEventMock } from '@mgdis/stencil-helpers';

const getPage = async (args, content = 'Text button') =>
  newSpecPage({
    components: [MgButton, MgForm],
    template: () => {
      const TagName = args.formTag;
      delete args.formTag;
      const button = () => <mg-button {...args}>{content}</mg-button>;
      if (!TagName) return button();
      else return <TagName identifier="identifier">{button()}</TagName>;
    },
  });

describe('mg-button', () => {
  beforeEach(() => {
    setupMutationObserverMock({
      observe: function () {
        return null;
      },
      disconnect: function () {
        return null;
      },
      takeRecords: () => [],
    });
    setupSubmitEventMock();
  });

  describe.each(['', ...variants])('Should render an %s button', variant => {
    test.each([false, true])('isIcon %s', async isIcon => {
      const { root } = await getPage({ variant, isIcon, label: 'label' });
      expect(root).toMatchSnapshot();
    });
  });

  test.each(buttonTypes)('Should render a button, case type %s', async type => {
    const { root } = await getPage({ label: 'label', type });
    expect(root).toMatchSnapshot();
  });

  test.each([true, false, undefined])('Should render a button, case full-width %s', async fullWidth => {
    const { root } = await getPage({ label: 'label', fullWidth });
    expect(root).toMatchSnapshot();
  });

  test('Should replace classes on variant changes', async () => {
    const page = await getPage({ variant: 'primary', label: 'label' });
    const element = page.doc.querySelector('mg-button');
    let classPrimary = element.shadowRoot.querySelector('.mg-c-button--primary');

    expect(classPrimary).not.toBeNull();

    // Change variant
    element.variant = 'danger';
    await page.waitForChanges();

    classPrimary = element.shadowRoot.querySelector('.mg-c-button--primary');
    const classDanger = element.shadowRoot.querySelector('.mg-c-button--danger');

    expect(classPrimary).toBeNull();
    expect(classDanger).not.toBeNull();
  });

  test.each([true, false])('Should update button on size to large with icon: %s', async isIcon => {
    const page = await getPage({ label: 'label', isIcon }, isIcon ? <mg-icon icon="user"></mg-icon> : undefined);
    const element = page.doc.querySelector('mg-button');

    expect(page.root).toMatchSnapshot();

    // Change size
    element.size = 'large';
    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
  });

  test.each([' ', 'blu'])('Should throw error with invalid variant: %s', async variant => {
    expect.assertions(1);
    try {
      await getPage({ variant, label: 'label' });
    } catch (err) {
      expect(err.message).toEqual(`<mg-button> prop "variant" must be one of: ${variants.join(', ')}. Passed value: ${variant}.`);
    }
  });

  test.each([' ', 'batman'])('Should throw error with invalid variant: %s', async size => {
    expect.assertions(1);
    try {
      await getPage({ size, label: 'label' });
    } catch (err) {
      expect(err.message).toEqual(`<mg-button> prop "size" must be one of: ${sizes.join(', ')}. Passed value: ${size}.`);
    }
  });

  test.each(['', ' ', undefined])('should throw error when using prop "isIcon" without a good prop "label"', async label => {
    expect.assertions(1);
    try {
      await getPage({ isIcon: true, label });
    } catch (err) {
      expect(err.message).toContain('<mg-button> prop "label" is mandatory when prop "isIcon" is set to true.');
    }
  });

  test('Should throw error when using prop "isIcon" without a good prop "fullWidth"', async () => {
    expect.assertions(1);
    try {
      await getPage({ isIcon: true, label: 'batman', fullWidth: true });
    } catch (err) {
      expect(err.message).toContain('<mg-button> prop "fullWidth" cannot be used with prop "isIcon"');
    }
  });

  describe('prevent double click', () => {
    test('Should NOT disable button after click', async () => {
      const page = await getPage({});
      const button = page.doc.querySelector('mg-button');
      const spy = jest.spyOn(page.rootInstance.disabledChange, 'emit');

      expect(page.root).toMatchSnapshot();

      button.dispatchEvent(new CustomEvent('click', { bubbles: true }));
      await page.waitForChanges();

      expect(spy).not.toHaveBeenCalled();
      expect(page.root).toMatchSnapshot();

      button.disabled = true;
      await page.waitForChanges();
      expect(spy).toHaveBeenCalledWith(true);

      expect(page.root).toMatchSnapshot();
    });

    test('Should disable button after click', async () => {
      const page = await getPage({ disableOnClick: true });
      const button = page.doc.querySelector('mg-button');
      const spy = jest.spyOn(page.rootInstance.disabledChange, 'emit');

      expect(page.root).toMatchSnapshot();

      button.dispatchEvent(new CustomEvent('click', { bubbles: true }));
      await page.waitForChanges();

      expect(spy).toHaveBeenCalledWith(true);
      expect(page.root).toMatchSnapshot();

      button.disabled = false;
      await page.waitForChanges();

      expect(spy).toHaveBeenCalledWith(false);
      expect(page.root).toMatchSnapshot();
    });

    test('Should not trigger disableOnClick when disabled', async () => {
      const page = await getPage({
        disabled: true,
        disableOnCLick: true,
      });
      const button = page.doc.querySelector('mg-button');
      const spy = jest.spyOn(page.rootInstance.disabledChange, 'emit');

      expect(page.root).toMatchSnapshot();

      button.dispatchEvent(new CustomEvent('click', { bubbles: true }));
      await page.waitForChanges();

      expect(spy).not.toHaveBeenCalled();
      expect(page.root).toMatchSnapshot();
    });

    test('Should not have fn when disabled', async () => {
      const page = await getPage({
        disabled: true,
        onClick: () => false,
      });

      expect(page.root).toMatchSnapshot();
    });
  });

  describe.each([{}, { disabled: true }])('keyboard', props => {
    test.each([' ', 'Enter', 'NumpadEnter'])('Should trigger click event on keydown', async key => {
      const page = await getPage(props);
      const button = page.doc.querySelector('mg-button');

      const spy = jest.spyOn(button, 'dispatchEvent');

      button.dispatchEvent(new KeyboardEvent(key === ' ' ? 'keyup' : 'keydown', { bubbles: true, key }));
      await page.waitForChanges();

      if (props.disabled) {
        expect(spy).not.lastCalledWith(expect.objectContaining({ type: 'click' }));
      } else {
        expect(spy).lastCalledWith(expect.objectContaining({ type: 'click' }));
      }
    });
    test.each(['ArrowRight', 'Tab', ' '])('Should NOT trigger click event on keydown', async key => {
      const page = await getPage(props);
      const button = page.doc.querySelector('mg-button');

      const spy = jest.spyOn(button, 'dispatchEvent');

      button.dispatchEvent(new KeyboardEvent('keydown', { bubbles: true, key }));
      await page.waitForChanges();

      expect(spy).not.lastCalledWith(expect.objectContaining({ type: 'click' }));
    });
  });

  describe.each(['form', 'mg-form'])('form <%s/>', form => {
    test.each([undefined, 'submit', 'button'])('Should emit "submit" event, case type is %s', async type => {
      const args = { type, formTag: form };
      const page = await getPage(args);

      const mgForm = page.doc.querySelector('mg-form')?.shadowRoot.querySelector('form') || page.doc.querySelector('form');
      const mgButton = page.doc.querySelector('mg-button');

      const formSpy = jest.spyOn(mgForm, 'dispatchEvent');

      mgButton.dispatchEvent(new Event('click', { bubbles: true }));

      await page.waitForChanges();

      if ([undefined, 'submit'].includes(type)) {
        expect(formSpy).toHaveBeenCalledWith(
          expect.objectContaining({
            type: 'submit',
            cancelable: true,
          }),
        );
      } else {
        expect(formSpy).not.toHaveBeenCalled();
      }
    });
  });
});
