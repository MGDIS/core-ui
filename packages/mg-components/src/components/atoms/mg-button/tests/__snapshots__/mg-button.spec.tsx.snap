// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mg-button Should render a button, case full-width false 1`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render a button, case full-width true 1`] = `
<mg-button aria-label="label" full-width="" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--full-width mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render a button, case full-width undefined 1`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render a button, case type button 1`] = `
<mg-button aria-label="label" role="button" tabindex="0" type="button">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render a button, case type reset 1`] = `
<mg-button aria-label="label" role="button" tabindex="0" type="reset">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render a button, case type submit 1`] = `
<mg-button aria-label="label" role="button" tabindex="0" type="submit">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render an  button isIcon false 1`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render an  button isIcon true 1`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--icon mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render an danger button isIcon false 1`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--danger mg-c-button--medium">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render an danger button isIcon true 1`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--danger mg-c-button--icon mg-c-button--medium">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render an danger-alt button isIcon false 1`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--danger-alt mg-c-button--medium">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render an danger-alt button isIcon true 1`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--danger-alt mg-c-button--icon mg-c-button--medium">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render an flat button isIcon false 1`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--flat mg-c-button--medium">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render an flat button isIcon true 1`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--flat mg-c-button--icon mg-c-button--medium">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render an info button isIcon false 1`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--info mg-c-button--medium">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render an info button isIcon true 1`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--icon mg-c-button--info mg-c-button--medium">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render an link button isIcon false 1`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--link mg-c-button--medium">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render an link button isIcon true 1`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--icon mg-c-button--link mg-c-button--medium">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render an primary button isIcon false 1`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render an primary button isIcon true 1`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--icon mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render an secondary button isIcon false 1`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--medium mg-c-button--secondary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render an secondary button isIcon true 1`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--icon mg-c-button--medium mg-c-button--secondary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render an success button isIcon false 1`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--medium mg-c-button--success">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render an success button isIcon true 1`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--icon mg-c-button--medium mg-c-button--success">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render tabindex attribute 1`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render tabindex attribute 2`] = `
<mg-button aria-label="label" role="button" tabindex="-1">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render tabindex attribute 3`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should render tabindex attribute 4`] = `
<mg-button aria-label="label" role="button" tabindex="1">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should update button on size to large with icon: false 1`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should update button on size to large with icon: false 2`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--large mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button Should update button on size to large with icon: true 1`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--icon mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  <mg-icon icon="user"></mg-icon>
</mg-button>
`;

exports[`mg-button Should update button on size to large with icon: true 2`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--icon mg-c-button--large mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  <mg-icon icon="user" size="large"></mg-icon>
</mg-button>
`;

exports[`mg-button Should update button on size to large with svg: %s 1`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--icon mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  <svg></svg>
</mg-button>
`;

exports[`mg-button Should update button on size to large with svg: %s 2`] = `
<mg-button aria-label="label" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--icon mg-c-button--large mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  <svg></svg>
</mg-button>
`;

exports[`mg-button Should update mg-icon size when button size is on default value 1`] = `
<mg-button aria-label="default size button" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  <mg-icon icon="user"></mg-icon>
</mg-button>
`;

exports[`mg-button prevent double click Should NOT disable button after click 1`] = `
<mg-button role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button prevent double click Should NOT disable button after click 2`] = `
<mg-button role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button prevent double click Should NOT disable button after click 3`] = `
<mg-button aria-disabled="true" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--disabled mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button prevent double click Should disable button after click 1`] = `
<mg-button role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button prevent double click Should disable button after click 2`] = `
<mg-button aria-disabled="true" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--disabled mg-c-button--loading mg-c-button--medium mg-c-button--primary">
      <mg-icon icon="loader" spin=""></mg-icon>
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button prevent double click Should disable button after click 3`] = `
<mg-button aria-disabled="false" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button prevent double click Should not have fn when disabled 1`] = `
<mg-button aria-disabled="true" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--disabled mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button prevent double click Should not trigger disableOnClick when disabled 1`] = `
<mg-button aria-disabled="true" disableonclick="" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--disabled mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;

exports[`mg-button prevent double click Should not trigger disableOnClick when disabled 2`] = `
<mg-button aria-disabled="true" disableonclick="" role="button" tabindex="0">
  <template shadowrootmode="open">
    <div class="mg-c-button mg-c-button--disabled mg-c-button--medium mg-c-button--primary">
      <div class="mg-c-button__content">
        <slot></slot>
      </div>
    </div>
  </template>
  Text button
</mg-button>
`;
